### PR TITLE
Failure Notification Example Via Slack

### PR DESCRIPTION
**Pull request by xplrior**
This example is to show how to set up a failure notification for a workflow.  This example sends a slack notification upon when the workflow exits with a failure. 

@lrochette this is for [Example: Notifications on Step/Pipeline Failure](https://github.com/codefresh-contrib/csdp01_git-source/issues/23)

### CODE REVIEW

File: examples/failure-notification/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Changelog
+
+## v0.0.1 (2022.09.30)
+
+### failure notifications
+
+This Workflow Template is an example of parsing and sending a slack notification on failure of a worklow.
```
Comment (lrochette) [inline examples/failure-notification/CHANGELOG.md:7]: workflow (last one) is misspelled online #7

File: examples/failure-notification/versions/0.0.1/docs/failure-notification.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
+# failure-notifications
```
Comment (lrochette) [inline examples/failure-notification/versions/0.0.1/docs/failure-notification.md:1]: singular to match template and file name

File: examples/failure-notification/versions/0.0.1/docs/failure-notification.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+## Summary
+
+This Workflow template is an example on how to parse and send a slack notification upon fialure of a workflow.
```
Comment (lrochette) [inline examples/failure-notification/versions/0.0.1/docs/failure-notification.md:5]: failure is misspelled

File: examples/failure-notification/versions/0.0.1/docs/failure-notification.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                - name: SLACK_TEXT
+                  value: "Workflow {{workflow.name}} has {{workflow.status}}. The template {{steps.parse-failure-info.outputs.parameters.templateName}} had the error of {{steps.parse-failure-info.outputs.parameters.message}} at {{steps.parse-failure-info.outputs.parameters.finishedAt}}"
+            when: "{{workflow.status}} != Succeeded"
+```
+
+## Example
+
+Workflow Failed with Exit Code 1.
+
+![workflow failure](../../../assets/workflow.png)
```
Comment (lrochette) [inline examples/failure-notification/versions/0.0.1/docs/failure-notification.md:86]: Link need to be absolute as they're displayed in https://codefresh.io/argohub
so 
https://raw.githubusercontent.com/codefresh-io/argo-hub//main/examples/failure-notification/assets/workflow.png

File: examples/failure-notification/versions/0.0.1/docs/failure-notification.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+Slack notification that gets sent.
+
+![workflow failure](../../../assets/slack.png)
```
Comment (lrochette) [inline examples/failure-notification/versions/0.0.1/docs/failure-notification.md:90]: same here
https://raw.githubusercontent.com/codefresh-io/argo-hub//main/examples/failure-notification/assets/slack.png

File: examples/failure-notification/versions/0.0.1/workflowTemplate.yaml [hunk 6]
```diff
@@ -0,0 +1,58 @@
+apiVersion: argoproj.io/v1alpha1
+kind: WorkflowTemplate
+metadata:
+  name: failure-notification-example
```
Comment (lrochette) [inline examples/failure-notification/versions/0.0.1/workflowTemplate.yaml:4]: name: argo-hub.failure-notification-example.0.0.1

File: examples/failure-notification/versions/0.0.1/workflowTemplate.yaml [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+spec:
```
Comment (lrochette) [inline examples/failure-notification/versions/0.0.1/workflowTemplate.yaml:5]: Need to add annotations aka (this is Matthew, adjust to your own for name a, URL and path)

```yaml
  annotations:
    argo-hub/version: '0.0.1'
    argo-hub/description: 'Workflow template for a full deployment using jira, sonar and a canary rollout'
    argo-hub/categories: 'argo'
    argo-hub/license: 'MIT'
    argo-hub/owner_name: 'Matthew Chung'
    argo-hub/owner_email: 'matthew.chung@codefresh.io'
    argo-hub/owner_avatar: 'https://github.com/matthewchungcodefresh.png'
    argo-hub/owner_url: 'https://github.com/matthewchungcodefresh'
    argo-hub/icon_url: "https://cdn.jsdelivr.net/gh/codefresh-io/argo-hub@main/examples/jira-sonar/assets/icon.svg"
    argo-hub/icon_background: "#f4f4f4"
```

File: examples/failure-notification/versions/0.0.1/workflowTemplate.yaml [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  entrypoint: intentional-fail
+  onExit: failure-notification # invoke pipeline-hook template at end of the workflow
+  templates:
+    # primary workflow template
+    - name: intentional-fail
+      container:
```
Comment (lrochette) [inline examples/failure-notification/versions/0.0.1/workflowTemplate.yaml:11]: Missing metadata

```yaml
      metadata:
        annotations:
          argo-hub-template/description: 'Builds a docker image using Kaniko, creates and updates a jira through each step, scans and tests the image repo, and deploys using a canary rollout'
          argo-hub-template/icon_url: "https://cdn.jsdelivr.net/gh/codefresh-io/argo-hub@main/examples/jira-sonar/assets/icon.svg"
          argo-hub-template/icon_background: "#f4f4f4"
```

File: examples/failure-notification/versions/0.0.1/workflowTemplate.yaml [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+              parameters:
+                - name: SLACK_HOOK_URL
+                  value: <INSERT SLACK URL>
+                - name: SLACK_TEXT
+                  value: "Workflow {{workflow.name}} has {{workflow.status}}. The template {{steps.parse-failure-info.outputs.parameters.templateName}} had the error of {{steps.parse-failure-info.outputs.parameters.message}} at {{steps.parse-failure-info.outputs.parameters.finishedAt}}"
+            when: "{{workflow.status}} != Succeeded"
+
+    # This Steps parses the information to be utlized in the slack notifcation.
+    - name: parse-info
+      serviceAccountName: codefresh-sa
```
Comment (lrochette) [inline examples/failure-notification/versions/0.0.1/workflowTemplate.yaml:39]: Missing annotations

**Review by lrochette — CHANGES_REQUESTED** [review]
I think the process is broken (at least was, @eti-codefresh ?), the build won't run after a merge from a fork. Only from a branch.
So You may have to merge first from your fork to a branch on codefresh-io, then from that branch to main.

for the workflow picture, use a Codefresh screenshot instead of the ArgoWorkflow one

**Comment by lrochette** [comment]
Maybe add at least 1  parameter for SLACK_HOOK_URL so it can be run as an example.

File: examples/failure-notification/CHANGELOG.md (hunk 1 above)
Comment (xplrior) [reply examples/failure-notification/CHANGELOG.md]: Fixed misspelling

File: examples/failure-notification/versions/0.0.1/docs/failure-notification.md (hunk 2 above)
Comment (xplrior) [reply examples/failure-notification/versions/0.0.1/docs/failure-notification.md]: fixed to show singular

File: examples/failure-notification/versions/0.0.1/docs/failure-notification.md (hunk 3 above)
Comment (xplrior) [reply examples/failure-notification/versions/0.0.1/docs/failure-notification.md]: Fixed misspelling

File: examples/failure-notification/versions/0.0.1/docs/failure-notification.md (hunk 4 above)
Comment (xplrior) [reply examples/failure-notification/versions/0.0.1/docs/failure-notification.md]: changed URL to be direct

File: examples/failure-notification/versions/0.0.1/docs/failure-notification.md (hunk 5 above)
Comment (xplrior) [reply examples/failure-notification/versions/0.0.1/docs/failure-notification.md]: changed URL to be direct

File: examples/failure-notification/versions/0.0.1/workflowTemplate.yaml (hunk 6 above)
Comment (xplrior) [reply examples/failure-notification/versions/0.0.1/workflowTemplate.yaml]: renamed workflow

File: examples/failure-notification/versions/0.0.1/workflowTemplate.yaml (hunk 9 above)
Comment (xplrior) [reply examples/failure-notification/versions/0.0.1/workflowTemplate.yaml:68]: added annotations

File: examples/failure-notification/versions/0.0.1/workflowTemplate.yaml (hunk 8 above)
Comment (xplrior) [reply examples/failure-notification/versions/0.0.1/workflowTemplate.yaml:30]: added annotations.

**Comment by xplrior** [comment]
@lrochette I have added the annotations, renamed the template, fixed the misspellings, added parameter, and retook a new screenshot.

**Comment by xplrior** [comment]
Closing PR as this is from fork.   Reopening new PR from branch.

**Pull request closed without merging**